go/libraries/doltcore/sqle/dtables/workspace.go

```diff
@@ -27,56 +27,368 @@ import (
+	if wtu.tableWriter != nil {
+		err := (*wtu.tableWriter).DiscardChanges(ctx, errorEncountered)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (wtu *WorkspaceTableUpdater) StatementComplete(ctx *sql.Context) error {
+	if wtu.err != nil {
+		return *wtu.err
+	}
+
+	return wtu.statementComplete(ctx)
+}
+
+func (wtm *WorkspaceTableModifier) statementComplete(ctx *sql.Context) error {
+	if wtm.tableWriter != nil {
+		err := (*wtm.tableWriter).Close(ctx)
+		if err != nil {
+			return err
+		}
+		wtm.tableWriter = nil
+	}
+
+	if wtm.sessionWriter != nil {
+		newWorkingSet, err := (*wtm.sessionWriter).Flush(ctx)
+		if err != nil {
+			return err
+		}
+		wtm.sessionWriter = nil
+
+		ds := dsess.DSessFromSess(ctx.Session)
+		err = ds.SetWorkingSet(ctx, ctx.GetCurrentDatabase(), newWorkingSet)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (wtu *WorkspaceTableUpdater) Update(ctx *sql.Context, old sql.Row, new sql.Row) error {
+	if old == nil || new == nil {
+		return fmt.Errorf("Runtime error: expected non-nil inputs to WorkspaceTableUpdater.Update")
+	}
+
+	valid, isStaged := validateWorkspaceUpdate(old, new)
+	if !valid {
+		return errors.New("only update of column 'staged' is allowed")
+	}
+
+	schemaLen := wtu.schemaLen
+
+	// old and new are the same. Just use one.
+	new = nil
+
+	toRow := old[3 : 3+schemaLen]
+	fromRow := old[3+schemaLen:]
+	if !isStaged {
+		toRow, fromRow = fromRow, toRow
+	}
+
+	// It's a delete if all the values in toRow are nil.
+	isDelete := true
+	for _, val := range toRow {
+		if val != nil {
+			isDelete = false
+			break
+		}
+	}
+
+	tableWriter := (*wtu.tableWriter)
+	if tableWriter == nil {
+		return fmt.Errorf("Runtime error: table writer is nil")
+	}
+
+	if isDelete {
+		return tableWriter.Delete(ctx, fromRow)
+	} else {
+		return tableWriter.Update(ctx, fromRow, toRow)
+	}
+}
+
+func (wtu *WorkspaceTableUpdater) Close(c *sql.Context) error {
+	// Resources released in StatementComplete
+	return nil
+}
+
+func (wtd *WorkspaceTableDeleter) StatementBegin(ctx *sql.Context) {
+	// Deletes are only allowed on WORKING, do not target staging.
+	sessionWriter, tableWriter, err := wtd.getWorkspaceTableWriter(ctx, false)
+	if err != nil {
+		wtd.err = &err
+		return
+	}
+	wtd.tableWriter = &tableWriter
+	wtd.sessionWriter = &sessionWriter
+}
+
+func (wtd *WorkspaceTableDeleter) DiscardChanges(ctx *sql.Context, errorEncountered error) error {
+	if wtd.tableWriter != nil {
+		err := (*wtd.tableWriter).DiscardChanges(ctx, errorEncountered)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (wtd *WorkspaceTableDeleter) StatementComplete(ctx *sql.Context) error {
+	if wtd.err != nil {
+		return *wtd.err
+	}
+
+	return wtd.statementComplete(ctx)
+}
+
+func (wtd *WorkspaceTableDeleter) Delete(c *sql.Context, row sql.Row) error {
+	isStaged := isTrue(row[stagedColumnIdx])
+	if isStaged {
+		return fmt.Errorf("cannot delete staged rows from workspace")
```

Comment:
Why not allow people to delete `staged=true` rows, too? It seems intuitive to do that when we allow non-staged rows to be deleted. Otherwise, they've got to figure out that they need to do an update to set `staged` to `false` and then run the delete. 

It seems handy and intuitive for both staged and unstaged rows to be able to delete them from this table to completely remove a change from the workspace and reset it back to its `HEAD` value.

Reply:
Discussed in person. Awkward to do this based on the layout of code. We are going to wait for a customer to ask for this before making changes. At the root of the issue is needing to modify two tables at the same time. Not impossible, just gross.